codec-http2/src/main/java/io/netty/handler/codec/http2/internal/hpack/Decoder.java

```diff
@@ -429,23 +404,113 @@ private static IllegalArgumentException notEnoughDataException(ByteBuf in) {
     }
 
     // Unsigned Little Endian Base 128 Variable-Length Integer Encoding
-    private static int decodeULE128(ByteBuf in) throws Http2Exception {
-        final int writerIndex = in.writerIndex();
-        for (int readerIndex = in.readerIndex(), shift = 0, result = 0;
-             readerIndex < writerIndex; ++readerIndex, shift += 7) {
-            byte b = in.getByte(readerIndex);
-            if (shift == 28 && (b & 0xF8) != 0) {
-                in.readerIndex(readerIndex + 1);
-                break;
-            }
-
-            if ((b & 0x80) == 0) {
+    private static int decodeULE128(ByteBuf in, int result) throws Http2Exception {
+        assert result <= 0x7f && result >= 0;
+        final int readerIndex = in.readerIndex();
+        byte b;
+
+        switch (in.readableBytes()) {
+            case 0:
+                // At this point we have already decoded a byte which requires more bytes to complete the int.
+                throw DECODE_ULE_128_DECOMPRESSION_EXCEPTION;
+            case 1:
+                b = in.getByte(readerIndex);
                 in.readerIndex(readerIndex + 1);
-                return result | ((b & 0x7F) << shift);
-            }
-            result |= (b & 0x7F) << shift;
+                if ((b & 0x80) != 0) {
+                    throw DECODE_ULE_128_DECOMPRESSION_EXCEPTION;
+                }
+                return result + (b & 0x7F);
+            case 2:
+                b = in.getByte(readerIndex);
+                result += b & 0x7F;
+                if ((b & 0x80) == 0) {
+                    in.readerIndex(readerIndex + 1);
+                    return result;
+                }
+                b = in.getByte(readerIndex + 1);
+                in.readerIndex(readerIndex + 2);
```

Comment:
Are you trying to avoid some extra processing done by `readByte`? It would seem that `readByte` would be effectively equivalent to:

```java
b = in.getByte(ix);
in.readerIndex(ix + 1);
```

no?